core/src/main/java/org/elasticsearch/index/Index.java

```diff
@@ -96,4 +110,37 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeString(name);
         out.writeString(uuid);
     }
+
+    @Override
+    public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startObject();
+        builder.field(INDEX_NAME_KEY, name);
+        builder.field(INDEX_UUID_KEY, uuid);
+        return builder.endObject();
+    }
+
+    public static Index fromXContent(final XContentParser parser) throws IOException {
+        return INDEX_PARSER.parse(parser, new Builder()).build();
+    }
+
+    /**
+     * Builder for Index objects.  Used by ObjectParser instances only,
+     * otherwise just call the constructor directly.
+     */
+    final public static class Builder {
```

Comment:
I think it can be done like this:

```
Index: core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java	(date 1459873883000)
+++ core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java	(revision )
@@ -37,6 +37,7 @@
 import java.util.EnumSet;
 import java.util.List;
 import java.util.Objects;
+import java.util.concurrent.TimeUnit;
 import java.util.function.BiFunction;
 
 /**
@@ -290,12 +291,14 @@
     final public static class Tombstone implements ToXContent, Writeable<Tombstone> {
 
         private static final String INDEX_KEY = "index";
-        private static final String DELETE_DATE_KEY = "delete_date_in_millis";
+        private static final String DELETE_DATE_IN_MILLIS_KEY = "delete_date_in_millis";
+        private static final String DELETE_DATE_IN_KEY = "delete_date";
         private static final ObjectParser<Tombstone.Builder, Void> TOMBSTONE_PARSER =
             new ObjectParser<>("tombstoneEntry", Tombstone.Builder::new);
         static {
-            TOMBSTONE_PARSER.declareObject(Tombstone.Builder::index, Index.INDEX_PARSER, new ParseField(INDEX_KEY));
-            TOMBSTONE_PARSER.declareLong(Tombstone.Builder::deleteDateInMillis, new ParseField(DELETE_DATE_KEY));
+            TOMBSTONE_PARSER.declareObject(Tombstone.Builder::index, Index::parseIndex, new ParseField(INDEX_KEY));
+            TOMBSTONE_PARSER.declareLong(Tombstone.Builder::deleteDateInMillis, new ParseField(DELETE_DATE_IN_MILLIS_KEY));
+            TOMBSTONE_PARSER.declareString((b,s) -> {}, new ParseField(DELETE_DATE_IN_KEY));
         }
 
         static BiFunction<XContentParser, Void, Tombstone> getParser() {
@@ -381,7 +384,7 @@
             builder.startObject();
             builder.field(INDEX_KEY);
             index.toXContent(builder, params);
-            builder.field(DELETE_DATE_KEY, deleteDateInMillis);
+            builder.timeValueField(DELETE_DATE_IN_MILLIS_KEY, DELETE_DATE_IN_KEY, deleteDateInMillis, TimeUnit.MILLISECONDS);
             return builder.endObject();
         }
 
@@ -393,10 +396,10 @@
          * A builder for building tombstone entries.
          */
         final private static class Builder {
-            private Index.Builder index;
+            private Index index;
             private long deleteDateInMillis = -1L;
 
-            public void index(final Index.Builder index) {
+            public void index(final Index index) {
                 this.index = index;
             }
 
@@ -407,7 +410,7 @@
             public Tombstone build() {
                 assert index != null;
                 assert deleteDateInMillis > -1L;
-                return new Tombstone(index.build(), deleteDateInMillis);
+                return new Tombstone(index, deleteDateInMillis);
             }
         }
     }
Index: core/src/main/java/org/elasticsearch/index/Index.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/src/main/java/org/elasticsearch/index/Index.java	(date 1459873883000)
+++ core/src/main/java/org/elasticsearch/index/Index.java	(revision )
@@ -40,12 +40,16 @@
     public static final Index[] EMPTY_ARRAY = new Index[0];
     private static final String INDEX_UUID_KEY = "index_uuid";
     private static final String INDEX_NAME_KEY = "index_name";
-    public static final ObjectParser<Builder, Void> INDEX_PARSER = new ObjectParser<>("index", Index.Builder::new);
+    private static final ObjectParser<Builder, Void> INDEX_PARSER = new ObjectParser<>("index", Index.Builder::new);
     static {
         INDEX_PARSER.declareString(Builder::name, new ParseField(INDEX_NAME_KEY));
         INDEX_PARSER.declareString(Builder::uuid, new ParseField(INDEX_UUID_KEY));
     }
 
+    public static final Index parseIndex(XContentParser parser, Void context) {
+        return INDEX_PARSER.apply(parser, context).build();
+    }
+
     private final String name;
     private final String uuid;
 
@@ -127,7 +131,7 @@
      * Builder for Index objects.  Used by ObjectParser instances only,
      * otherwise just call the constructor directly.
      */
-    final public static class Builder {
+    final private static class Builder {
         private String name;
         private String uuid;

```